### ISSUE TITLE
[Blocker] Build breakage at 3.0.3 on NonStop

### ISSUE THREAD

**Issue by rsbeckerca** [openssl/openssl#18232]
As of 3.0.3, NonStop no longer builds. `fipsprov.c` now uses `locale_t` which was not in the previous version. Builds no longer work, as follows:

```
c99: running /usr/cmplr/ccomfe -Z8 -ZBL --c99 -ZO2 -Zg3 -ZS/tmp/t134217835AFDIHB.T -D_XOPEN_SOURCE -ZR8 -m -Zf -ZT -D_TANDEM_SOURCE -ZNW203 -ZNW220 -ZNW272 -ZNW734 -ZNW770 -ZNW1506 -Zz -I. -Icrypto -Iinclude -Iproviders/implementations/include -Iproviders/common/include -I/usr/include -DFIPS_MODULE -DOPENSSL_PIC -DOPENSSLDIR="/usr/local-ssl3.0/ssl" -DENGINESDIR="/usr/local-ssl3.0/lib64/engines-3" -DMODULESDIR="/usr/local-ssl3.0/lib64/ossl-modules" -DOPENSSL_BUILDING_OPENSSL -DOPENSSL_VPROC= -D_XOPEN_SOURCE -D_XOPEN_SOURCE_EXTENDED=1 -D_TANDEM_SOURCE -DB_ENDIAN -D_TANDEM_ARCH=3 -DNDEBUG --rvu_version=1605 providers/fips/fipsprov.c 
  static locale_t loc;
         ^
"/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/providers/fips/fipsprov.c", line 50: error(114):
          identifier "locale_t" is undefined

      loc = newlocale(LC_COLLATE_MASK, "C", (locale_t) 0);
            ^
"/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/providers/fips/fipsprov.c", line 514: error(114):
          identifier "newlocale" is undefined

      loc = newlocale(LC_COLLATE_MASK, "C", (locale_t) 0);
                      ^
"/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/providers/fips/fipsprov.c", line 514: error(114):
          identifier "LC_COLLATE_MASK" is undefined

      loc = newlocale(LC_COLLATE_MASK, "C", (locale_t) 0);
                                             ^
"/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/providers/fips/fipsprov.c", line 514: error(114):
          identifier "locale_t" is undefined

      loc = newlocale(LC_COLLATE_MASK, "C", (locale_t) 0);
                                                       ^
"/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/providers/fips/fipsprov.c", line 514: error(112):
          expected a ")"

      return (loc == (locale_t) 0) ? 0 : 1;
                                ^
"/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/providers/fips/fipsprov.c", line 516: error(112):
          expected a ")"

      freelocale(loc);
      ^
"/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/providers/fips/fipsprov.c", line 520: error(114):
          identifier "freelocale" is undefined
```

**Comment by beldmit** [comment]
Could you please adjust this file https://github.com/openssl/openssl/blob/master/crypto/ctype.c#L299-L304 for your platform?

**Comment by beldmit** [comment]
I'm sorry for breaking your build

**Comment by rsbeckerca** [comment]
> Could you please adjust this file https://github.com/openssl/openssl/blob/master/crypto/ctype.c#L299-L304 for your platform?

I am not sure what you want me to adjust. `locale.h` on the platform does not contain `locale_t`, `newlocale`, `freelocale`, `LC_COLLATE_MASK`. These definitions appear to be in POSIX as of 2018 and are not available broadly.

**Comment by beldmit** [comment]
On Windows there are similar functions with slightly different names. On Mac we need to include a different header file. And according to manuals, it should be POSIX 2008, not 2018.

Can I help you somehow with these adjustments?

**Comment by rsbeckerca** [comment]
The point is, this is a new API usage in the middle of a fix series. There is no way to adjust header files to include definitions on the platform that do not exist.

**Comment by beldmit** [comment]
I understand your point, and it's a real problem that fixing the issue required a new API.

You still can adjust the defines to fall back to old implementation: https://github.com/openssl/openssl/blob/master/crypto/ctype.c#L297

**Comment by rsbeckerca** [comment]
> I understand your point, and it's a real problem that fixing the issue required a new API.
> 
> You still can adjust the defines to fall back to old implementation: https://github.com/openssl/openssl/blob/master/crypto/ctype.c#L297

I still do not get how to fix this and still have a valid 3.0.3 build? What issue was fixed that had to be in 3.0.3 and I still don't see how I can revert - I'm not the fix author here and we have limited locale support.

**Comment by beldmit** [comment]
What are your platform-specific defines?

**Comment by rsbeckerca** [comment]
The compiler defines `__TANDEM`. OpenSSL `Configurations/50-nonstop.cnf` defines `_TANDEM_SOURCE`. The `sys_id` is `TANDEM`

**Comment by beldmit** [comment]
So could you please try
```
#if ((defined OPENSSL_SYS_WINDOWS) || (defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 200809L)) && !defined (__TANDEM)
```

**Comment by rsbeckerca** [comment]
Made no difference:
```
  static locale_t loc;
         ^
"/home/ituglib/randall/openssl-3.0/providers/fips/fipsprov.c", line 50: error(114):
          identifier "locale_t" is undefined

      loc = newlocale(LC_COLLATE_MASK, "C", (locale_t) 0);
            ^
"/home/ituglib/randall/openssl-3.0/providers/fips/fipsprov.c", line 514: error(114):
          identifier "newlocale" is undefined

      loc = newlocale(LC_COLLATE_MASK, "C", (locale_t) 0);
                      ^
"/home/ituglib/randall/openssl-3.0/providers/fips/fipsprov.c", line 514: error(114):
          identifier "LC_COLLATE_MASK" is undefined

      loc = newlocale(LC_COLLATE_MASK, "C", (locale_t) 0);
                                             ^
"/home/ituglib/randall/openssl-3.0/providers/fips/fipsprov.c", line 514: error(114):
          identifier "locale_t" is undefined

      loc = newlocale(LC_COLLATE_MASK, "C", (locale_t) 0);
                                                       ^
"/home/ituglib/randall/openssl-3.0/providers/fips/fipsprov.c", line 514: error(112):
          expected a ")"

      return (loc == (locale_t) 0) ? 0 : 1;
                                ^
"/home/ituglib/randall/openssl-3.0/providers/fips/fipsprov.c", line 516: error(112):
          expected a ")"

      freelocale(loc);
      ^
"/home/ituglib/randall/openssl-3.0/providers/fips/fipsprov.c", line 520: error(114):
          identifier "freelocale" is undefined

```

**Comment by rsbeckerca** [comment]
I am trying to understand why standard `strcasecmp` and strncasecmp` are not acceptable default implementations for platforms that do not support `locale_t`. If this was done at 3.1, we might have had a chance at providing an alternate implementation. Now, we have a blocker.

**Comment by beldmit** [comment]
Please try this patch.

[tandem.txt](https://github.com/openssl/openssl/files/8613820/tandem.txt)

I'm sorry, when I tried to ping you for test, I used a wrong incarnation (@rsbecker)

**Comment by rsbeckerca** [comment]
I do have an _LC_locale_t on the platform. It looks like the locale definitions are based on `OSF/1 1.3 (Revision 3.0.5.2)`. There are no new/free calls - those are `libc` definitions.

**Comment by beldmit** [comment]
If you use it and provide proper defines similar to Windows ones, it would definitely be better.

**Comment by rsbeckerca** [comment]
> If you use it and provide proper defines similar to Windows ones, it would definitely be better.

I cannot get into the underlying library, so not really an option.

**Comment by beldmit** [comment]
So then this patch together with this change https://github.com/openssl/openssl/issues/18232#issuecomment-1116423127 should fix most of problems (though, of cause, may need some tweaks)

**Comment by rsbeckerca** [comment]
If I had access to the operating system source code, many other problems would go away. Working on the patch.

**Comment by rsbeckerca** [comment]
PR created for this. It is WIP until I get the tests working.

**Comment by rsbeckerca** [comment]
I'm having issues with FIPS tests. Going to start with a clean clone and try again.

**Comment by rsbeckerca** [comment]
Tests now running - at 15-test_gendsa.t so far

**Comment by rsbeckerca** [comment]
Test passes on NonStop. Updating PR.

**Comment by florinspatar** [comment]
Same issue on HP-UX, AIX, Solaris and older Linux (RHEL5 & SLES10). Also tested with the proposed patch, but there is no difference.

**Comment by beldmit** [comment]
@florinspatar what do you mean by proposed patch? #18233 is the patch but it should be tuned to provide other platform-specific #ifndef

**Comment by florinspatar** [comment]
I used patch from #18233. I understand that it should be adapted for other platforms as well, before it can work as expected.

**Comment by rsbeckerca** [comment]
> Same issue on HP-UX, AIX, Solaris and older Linux (RHEL5 & SLES10). Also tested with the proposed patch, but there is no difference.

It sounds like the breakage I reported in this case goes far beyond just NonStop. This is why dependency changes should not be done within a series (a.k.a. 3.0.x) but should be deferred to 3.1. If the Pull Request needs to account for other platforms, I can make the changes but am going to need a list of the system defines. This has, unsurprisingly, become a much bigger issue that originally reported, making 3.0.3 essentially DOA.

**Comment by t8m** [comment]
> This is why dependency changes should not be done within a series (a.k.a. 3.0.x) but should be deferred to 3.1.

There is no way around fixing the strcasecmp problem other than doing what we did. Apart from doing our own implementation which would have its own set of serious problems like performance regression.

> This has, unsurprisingly, become a much bigger issue that originally reported, making 3.0.3 essentially DOA.

Only for quite marginal systems which are not our primary targets.

Users of any system that are not within our primary targets should test the current code on the branches they use before release to report regressions. Or they should be prepared to patch the source post release.

**Comment by rsbeckerca** [comment]
> > I agree, but I did not get the code until the day 3.0.3 was released. Perhaps the platform/community maintainers can get the code a day or two early to catch situations like this - the x86 build/test takes about 2-3 hours so we could have caught this before general release. Because we only got this day of, we end up working multiple 24 hours straight to try to catch up and deliver the compiled packages to our base. It's exhausting and stressful.
> 
> Is it feasible to setup a buildbot worker on your system? Ultimately it would be good to automate the build/test process. In this particular case the problematic code was in the tree for a while before the actual 3.0.3 release. If we could hook your system into CI then we could have caught it earlier. (Note: our buildbot infrastructure is a "work in progress", but we have it working for some internally hosted systems already....eventually we want to integrate it into github, but we're not quite there yet).
> 
> That doesn't necessarily help with "private" security fixes - but the problem here wasn't with one of those issues.

We have a fully automated build/test system for 3.0.x. It currently depends on content showing up on the openssl-3.0 branch, which gets polled roughly every two days. We cannot have GitHub trigger anything externally as our team does not control the VPN structure - and there is no compatible VPN for the GitHub servers. Should we have seen the changes, because we did not. We did find the master breakages (there are two test breakages that have been unresolved for months) this way.

**Comment by mattcaswell** [comment]
> We have a fully automated build/test system for 3.0.x. It currently depends on content showing up on the openssl-3.0 branch, which gets polled roughly every two days. We cannot have GitHub trigger anything externally as our team does not control the VPN structure - and there is no compatible VPN for the GitHub servers. Should we have seen the changes, because we did not.

Yes - you should have seen these changes. They were introduced to the 3.0 branch by #18103 on 21st April - so a little under 2 weeks before the 3.0.3 release.

For buildbot you would not be communicating with GitHub - but rather our own buildbot infrastructure. I am not a buildbot expert (@levitte is), but IIUC no connection would be made from our infrastructure to yours. Rather it is the other way around, i.e. you periodically connect to our buildbot master. I don't know if that make it any easier for you.

**Comment by rsbeckerca** [comment]
> > We have a fully automated build/test system for 3.0.x. It currently depends on content showing up on the openssl-3.0 branch, which gets polled roughly every two days. We cannot have GitHub trigger anything externally as our team does not control the VPN structure - and there is no compatible VPN for the GitHub servers. Should we have seen the changes, because we did not.
> 
> Yes - you should have seen these changes. They were introduced to the 3.0 branch by #18103 on 21st April - so a little under 2 weeks before the 3.0.3 release.
> 
> For buildbot you would not be communicating with GitHub - but rather our own buildbot infrastructure. I am not a buildbot expert (@levitte is), but IIUC no connection would be made from our infrastructure to yours. Rather it is the other way around, i.e. you periodically connect to our buildbot master. I don't know if that make it any easier for you.

It is probably easier to just monitor git branches, providing openssl-3.0 is updated. I tried porting buildbot but it has components that are not available on NonStop (porting dependencies apparently on gcc, which we do not have). I will increase the poll frequency.

**Comment by rsbeckerca** [comment]
> > We have a fully automated build/test system for 3.0.x. It currently depends on content showing up on the openssl-3.0 branch, which gets polled roughly every two days. We cannot have GitHub trigger anything externally as our team does not control the VPN structure - and there is no compatible VPN for the GitHub servers. Should we have seen the changes, because we did not.
> 
> Yes - you should have seen these changes. They were introduced to the 3.0 branch by #18103 on 21st April - so a little under 2 weeks before the 3.0.3 release.
> 
> For buildbot you would not be communicating with GitHub - but rather our own buildbot infrastructure. I am not a buildbot expert (@levitte is), but IIUC no connection would be made from our infrastructure to yours. Rather it is the other way around, i.e. you periodically connect to our buildbot master. I don't know if that make it any easier for you.

@mattcaswell I have a monitor now for `master` and `openssl-3.0` branches. Currently both fail to build because of this particular issue. Hoping we get a merge of #18241 soon. This should catch breakages within a day or two - note that we know there are at least two test failures still pending on `master` relating to mixing the same code in `.so` and `.a`.

**Comment by Tarnum-tst** [comment]
I got buid failure here too on my old Windows system with mingw32 (w/o locale support). To fix it I restored two removed lines in include/internal/e_os.h:
#  define strcasecmp _stricmp
#  define strncasecmp _strnicmp
and replaced this line in crypto/ctype.c
#if (defined OPENSSL_SYS_WINDOWS) || (defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 200809L)
with
#if 0

**Comment by rsbeckerca** [comment]
@Tarnum-tst If I am not mistaken, I think that basically reverts the change.

**Comment by t8m** [comment]
Fixed by #18293
